### PR TITLE
*: remove varsutil package, make Systems a private member of SessionVars

### PR DESCRIPTION
**Pull request by tiancaiamao**
Session system variable may bind to values in `SessionVars`, change `SessionVars.Systems` directly makes things inconsistent, so I make `Systems` a private member and wrap get/set methods around it.

@shenli @coocood

### CODE REVIEW

File: sessionctx/varsutil/varsutil_test.go [hunk 1]
```diff
@@ -34,28 +34,6 @@ var _ = Suite(&testVarsutilSuite{})
 type testVarsutilSuite struct {
 }
 
-func (s *testVarsutilSuite) TestTiDBOptOn(c *C) {
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply sessionctx/varsutil/varsutil_test.go]: Now I add it back.

**Comment by coocood** [comment]
@tiancaiamao 
I think we can remove `varsutil` package.

**Comment by tiancaiamao** [comment]
Agree. Should I do it in this PR or in a new one? @coocood

**Comment by coocood** [comment]
@tiancaiamao prefer this one.

**Comment by shenli** [comment]
@tiancaiamao Please resolve the conflicts.

File: sessionctx/variable/varsutil.go [hunk 2]
```diff
@@ -99,85 +98,24 @@ func GetScopeNoneSystemVar(key string) (string, bool, error) {
...
-		vars.BatchDelete = tidbOptOn(sVal)
-	case variable.TiDBDMLBatchSize:
-		vars.DMLBatchSize = tidbOptPositiveInt(sVal, variable.DefDMLBatchSize)
-	case variable.TiDBCurrentTS:
-		return variable.ErrReadOnly
-	case variable.TiDBMaxChunkSize:
-		vars.MaxChunkSize = tidbOptPositiveInt(sVal, variable.DefMaxChunkSize)
-	case variable.TiDBMemThreshold:
-		vars.MemThreshold = int64(tidbOptPositiveInt(sVal, variable.DefMemThreshold))
-	case variable.TiDBGeneralLog:
```
Comment (coocood) [inline sessionctx/variable/varsutil.go]: GenaralLog is missed.

File: sessionctx/variable/varsutil.go [hunk 3]
```diff
@@ -99,85 +98,24 @@ func GetScopeNoneSystemVar(key string) (string, bool, error) {
...
-		vars.BatchInsert = tidbOptOn(sVal)
-	case variable.TiDBBatchDelete:
-		vars.BatchDelete = tidbOptOn(sVal)
-	case variable.TiDBDMLBatchSize:
-		vars.DMLBatchSize = tidbOptPositiveInt(sVal, variable.DefDMLBatchSize)
-	case variable.TiDBCurrentTS:
-		return variable.ErrReadOnly
-	case variable.TiDBMaxChunkSize:
-		vars.MaxChunkSize = tidbOptPositiveInt(sVal, variable.DefMaxChunkSize)
-	case variable.TiDBMemThreshold:
```
Comment (coocood) [inline sessionctx/variable/varsutil.go]: MemThreshold is missed.

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
PTAL @zimulala

File: sessionctx/variable/session.go [hunk 4]
```diff
@@ -140,8 +142,8 @@ type SessionVars struct {
 	UsersLock sync.RWMutex
 	// Users are user defined variables.
 	Users map[string]string
-	// Systems are system variables.
-	Systems map[string]string
+	// system variables, don't modify it directly, use GetSystemVar/SetSystemVar method.
```
Comment (zimulala) [inline sessionctx/variable/session.go]: The format needs to like `systems ....`

File: sessionctx/variable/session.go [hunk 5]
```diff
@@ -394,6 +396,88 @@ func (s *SessionVars) ResetPrevAffectedRows() {
...
 }
 
+// GetSystemVar gets the string value of a system variable.
+func (s *SessionVars) GetSystemVar(name string) (string, bool) {
+	val, ok := s.systems[name]
+	return val, ok
+}
+
+// DeleteSystemVar deletes a system variable.
+func (s *SessionVars) DeleteSystemVar(name string) error {
```
Comment (zimulala) [inline sessionctx/variable/session.go]: This function doesn't be used. Do we need it?

Comment (zimulala) [reply sessionctx/variable/session.go]: Can we use it on line 109 in the file of sessionctx/variable/varsutil.go

File: executor/adapter.go [hunk 6]
```diff
@@ -197,19 +197,19 @@ func (a *ExecStmt) Exec(goCtx goctx.Context) (ast.RecordSet, error) {
...
 	ctx := a.Ctx
 	if _, ok := a.Plan.(*plan.Analyze); ok && ctx.GetSessionVars().InRestrictedSQL {
-		oriStats := ctx.GetSessionVars().Systems[variable.TiDBBuildStatsConcurrency]
+		oriStats, _ := ctx.GetSessionVars().GetSystemVar(variable.TiDBBuildStatsConcurrency)
 		oriScan := ctx.GetSessionVars().DistSQLScanConcurrency
 		oriIndex := ctx.GetSessionVars().IndexSerialScanConcurrency
-		oriIso := ctx.GetSessionVars().Systems[variable.TxnIsolation]
-		ctx.GetSessionVars().Systems[variable.TiDBBuildStatsConcurrency] = "1"
+		oriIso, _ := ctx.GetSessionVars().GetSystemVar(variable.TxnIsolation)
+		terror.Log(ctx.GetSessionVars().SetSystemVar(variable.TiDBBuildStatsConcurrency, "1"))
```
Comment (zz-jason) [inline executor/adapter.go]: we should `terror.Log(errors.Trace(...))`

**Comment by zimulala** [comment]
LGTM

**Pull request merged**